src/covid_model_seiir_pipeline/ode_fit/model.py

```diff
@@ -0,0 +1,389 @@
+
+        if np.any(neg_S_idx):
+            id_min = np.min(np.arange(S.size)[neg_S_idx])
+            S[id_min:] = S[id_min - 1]
+            E[id_min:] = E[id_min - 1]
+            I1[id_min:] = I1[id_min - 1]
+            I2[id_min:] = I2[id_min - 1]
+            R[id_min:] = R[id_min - 1]
+
+        self.components = {
+            'S': S,
+            'E': E,
+            'I1': I1,
+            'I2': I2,
+            'R': R
+        }
+
+        # get beta
+        self.params = (self.rhs_newE/
+                       ((S/self.N)*
+                        ((I1 + I2)**self.alpha)))[None, :]
+
+    def predict(self, t):
+        params = linear_interpolate(t, self.t_params, self.params)
+        components = {
+            c: linear_interpolate(t, self.t_params, self.components[c])
+            for c in self.components
+        }
+        components.update({
+            'newE': linear_interpolate(t, self.t_params, self.rhs_newE)
+        })
+        components.update({
+            'newE_obs': linear_interpolate(t, self.t, self.obs)
+        })
+        return params, components
+
+    def create_result_df(self):
+        """Create result DataFrame.
+        """
+        params, components = self.predict(self.t)
+        df_result = pd.DataFrame({
+            self.col_loc_id: self.loc_id,
+            self.col_date: self.date,
+            'days': self.t,
+            'beta': params[0]
+        })
+
+        for k, v in components.items():
+            df_result[k] = v
+
+        return df_result
+
+    def create_params_df(self):
+        """Create parameter DataFrame.
+        """
+        df_params = pd.DataFrame([self.alpha, self.sigma,
+                                  self.gamma1, self.gamma2],
+                                 index=['alpha', 'sigma', 'gamma1', 'gamma2'],
+                                 columns=['params'])
+
+        return df_params
+
+    def create_start_end_date_df(self):
+        df_result = pd.DataFrame({
+            self.col_loc_id: self.loc_id,
+            'start_date': self.start_date,
+            'end_date': self.end_date
+        }, index=[0])
+        return df_result
+
+
+@dataclass
+class ODEProcessInput:
+    df_dict: Dict
+    col_date: str
+    col_cases: str
+    col_pop: str
+    col_loc_id: str
+    col_lag_days: str
+    col_observed: str
+
+    alpha: Tuple
+    sigma: Tuple
+    gamma1: Tuple
+    gamma2: Tuple
+    solver_dt: float
+    day_shift: Tuple
+
+
+class ODEProcess:
+    """ODE Process for multiple group.
+    """
+    def __init__(self, input):
+        """Constructor of ODEProcess.
+
+        Args:
+            input: ODEProcessInput
+        """
+        self.df_dict = input.df_dict
+        self.col_date = input.col_date
+        self.col_cases = input.col_cases
+        self.col_pop = input.col_pop
+        self.col_loc_id = input.col_loc_id
+        self.col_lag_days = input.col_lag_days
+        self.col_observed = input.col_observed
```

Comment:
`df_dict` and `col_date` (and it appears the rest - but I only CTRL + F'd the first two) are only used within this function. Do they need to be member variables instead of just assigning `input` to `self`?

Reply:
They do not.  This whole module has a bunch of stuff that we'll want to clean up and simplify.  Right now the ODEProcessInput is a pretty useless struct.  It's packed, passed in, then immediately unpacked.